src/v/transform/rpc/deps.h

```diff
@@ -0,0 +1,123 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+#pragma once
+
+#include "cluster/fwd.h"
+#include "cluster/types.h"
+#include "kafka/server/partition_proxy.h"
+#include "model/fundamental.h"
+#include "model/ktp.h"
+#include "model/metadata.h"
+
+#include <seastar/util/noncopyable_function.hh>
+
+/**
+ * These are wrapper types for types defined outside this module, useful for
+ * testing without spinning up a full Redpanda cluster.
+ */
+
+namespace transform::rpc {
+
+/**
+ * A cache for which node owns a given partition leader.
+ */
+class partition_leader_cache {
+public:
+    partition_leader_cache() = default;
+    partition_leader_cache(const partition_leader_cache&) = delete;
+    partition_leader_cache(partition_leader_cache&&) = delete;
+    partition_leader_cache& operator=(const partition_leader_cache&) = delete;
+    partition_leader_cache& operator=(partition_leader_cache&&) = delete;
+    virtual ~partition_leader_cache() = default;
+
+    static std::unique_ptr<partition_leader_cache>
+    make_default(ss::sharded<cluster::partition_leaders_table>*);
+
+    /**
+     * Lookup which node owns a given ntp's leader.
+     */
+    std::optional<model::node_id> get_leader_node(const model::ntp& ntp) const;
+
+    virtual std::optional<model::node_id>
+      get_leader_node(model::topic_namespace_view, model::partition_id) const
+      = 0;
+};
+
+/**
+ * A cache for topic/partition metadata.
+ */
+class topic_metadata_cache {
+public:
+    topic_metadata_cache() = default;
+    topic_metadata_cache(const topic_metadata_cache&) = delete;
+    topic_metadata_cache(topic_metadata_cache&&) = delete;
+    topic_metadata_cache& operator=(const topic_metadata_cache&) = delete;
+    topic_metadata_cache& operator=(topic_metadata_cache&&) = delete;
+    virtual ~topic_metadata_cache() = default;
+
+    static std::unique_ptr<topic_metadata_cache>
+    make_default(ss::sharded<cluster::metadata_cache>*);
+
+    /**
+     * Lookup the topic configuration.
+     */
+    virtual std::optional<cluster::topic_configuration>
+      find_topic_cfg(model::topic_namespace_view) const = 0;
+
+    /**
+     * The default max batch bytes size.
+     */
+    virtual uint32_t get_default_batch_max_bytes() const = 0;
+};
+
+/**
+ * Handles routing for shard local partitions.
+ */
+class partition_manager {
```

Comment:
should we build interface types outside of transform::rpc so other subsystems don't need to repeat this?

Reply:
Yes I think so, or we should have a single interface type for the entire data path that the different entry points come into (kafka, internal, future protocol X). I would love to talk through the right place organizationally for this with you. I don't like how large the cluster target is and maybe we just need a "data path" target shared by this and the kafka server?

Reply:
sgtm. we can prototype in transform::* and then move it out later